File: src/simcore/flow.py
"""Flows: structured, atomic, stoichiometric transfers (decisions #1/#2).

A ``Flow`` exposes only ``evaluate`` — legs exist **only after** evaluation
against a snapshot (the key step-3 correction). So balance and referential
integrity are *evaluation-time* properties, not registration-time:

  * **Balance** is a property of an evaluated ``FlowResult``: per conserved
    ``Quantity``, the legs sum to ~0 (every mole withdrawn from one carbon stock
    lands in another carbon stock, boundary reservoirs included). Resolving a
    leg's quantity needs a stock lookup, so the check is a pure helper over
    ``(result, stocks)`` — used here and **reused by the step-8 every-step
    conservation gate** (hence the shared ``ASSERTED_QUANTITIES`` / ``BALANCE_*``
    contract living in ``simcore.quantities``).
  * **Referential integrity** (every produced ``Leg.stock`` is a real stock) is
    *not* checked here — it is an assertion in the apply path (step 5/6), which
    runs before the conservation gate. A typo'd stock id surfaces as a
    ``KeyError`` at the first step, by design.

``ENERGY`` is **now asserted** like the mass quantities — it joined the conserved
set in Phase 5 (the Power domain; was exempt through Phase 4 under the original
decision #8). ``per_quantity_residual`` reports every quantity present (it always
did); ``assert_flow_balanced`` iterates ``ASSERTED_QUANTITIES``, which now includes
``ENERGY``. Electricity→heat is a transmutation of *form* within the one ENERGY
currency (see ``simcore.quantities`` / phase-5 plan), not a cross-quantity flow.

Pure stdlib only.
"""

import math
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Protocol, runtime_checkable

from simcore.environment import Environment
from simcore.ids import DomainId, FlowId, StockId
from simcore.quantities import (
    ASSERTED_QUANTITIES,
    BALANCE_ATOL,
    BALANCE_RTOL,
    Quantity,
)
from simcore.state import State, Stock


class ConservationError(Exception):
    """A flow/ledger failed per-quantity balance within tolerance.

    Such a failure is an engine bug (decisions #2/#13), not a recoverable
    condition. It gets a dedicated type — not ``ValueError`` (brittle to
    message-match) and not ``AssertionError`` (stripped under ``python -O``) — so
    the step-8 every-step gate can catch/assert *this specific* failure.
    """


@dataclass(frozen=True)
class Leg:
    """One stock touched by a flow.

    ``amount`` is per dt in the stock's canonical unit: ``>0`` deposits into the
    stock, ``<0`` withdraws from it.
    """

    stock: StockId
    amount: float

    def __post_init__(self) -> None:
        # NaN/Inf are forbidden anywhere in state (determinism/serialization).
        if not math.isfinite(self.amount):
            raise ValueError(
                f"Leg on {self.stock!r} amount is not finite: {self.amount!r}"
            )


@dataclass(frozen=True)
class FlowResult:
    """The requested transfer from one ``evaluate`` against a snapshot.

    At most one leg per ``StockId`` — a flow nets its own touches on a stock into
    a single leg. This keeps arbitration's per-(stock, flow) ``demand``/``scale``
    a clean scalar (step 7). Empty ``legs`` is a valid no-op. ``legs`` is coerced
    to a tuple so the result stays hashable/comparable (the purity test compares
    two results for equality); the annotation stays ``tuple[Leg, ...]`` so the
    checker still flags non-tuples at call sites.
    """

    legs: tuple[Leg, ...]

    def __post_init__(self) -> None:
        legs = tuple(self.legs)
        object.__setattr__(self, "legs", legs)
        seen: set[StockId] = set()
        for leg in legs:
            if leg.stock in seen:
                raise ValueError(
                    f"FlowResult has a duplicate leg for stock {leg.stock!r}; "
                    "a flow must net its own touches on a stock into one leg"
                )
            seen.add(leg.stock)


@runtime_checkable
class Flow(Protocol):
    """A pure, deterministic stoichiometric transfer.

    ``evaluate`` reads the snapshot/env only, never mutates, and is deterministic
    in its inputs. ``priority`` is carried for declared-controller arbitration
    policies (decision #5) but is **unused** under the proportional default;
    canonical reduction order is always id-sorted (#15), never priority-sorted.

    **Increment-form contract (step 6).** ``evaluate`` returns the *per-step
    increment* — each leg amount is ``dt·rate(snapshot)``, not a bare rate — and
    ``rate`` must be **independent of ``dt``** (linear in ``dt``). This is what
    lets RK4's ⅙-combine reproduce classical RK4 exactly (see
    ``simcore.integrator``). A flow that uses ``dt`` non-linearly (an analytic
    sub-step, dt-gated logic) still conserves mass but silently forfeits RK4 order.

    ``id``/``priority`` are **read-only** (declared as properties) so that frozen
    flow implementations — the expected shape, matching the immutable-state model
    — satisfy the protocol; a mutable attribute satisfies a read-only member too.
    """

    @property
    def id(self) -> FlowId: ...

    @property
    def priority(self) -> int: ...

    def evaluate(self, snapshot: State, env: Environment, dt: float) -> FlowResult: ...


def per_quantity_residual(
    result: FlowResult, stocks: Mapping[StockId, Stock]
) -> dict[Quantity, float]:
    """Net leg sum per ``Quantity`` actually present in ``result`` (the diagnostic).

    Includes ``ENERGY`` (exempt from the asserted balance but reported as a
    diagnostic, decision #8). Quantities not touched are simply absent from the
    result (a missing quantity has a trivial 0 residual). Legs are folded in
    **canonical order** — sorted by stock id (decision #15) — so the sum is
    bit-identical regardless of leg construction order; this helper is the
    building block the step-8 every-step gate composes. (Phase-0 ids are ASCII,
    so Python's str sort agrees with the future Rust UTF-8 byte sort — keep ids
    ASCII.)

    A leg referencing an unknown stock id raises ``KeyError`` — referential
    integrity is the apply path's job (step 5/6), not this helper's.
    """
    residual: dict[Quantity, float] = {}
    for leg in sorted(result.legs, key=lambda leg: leg.stock):
        # Element-composition fold (P2.1): a leg contributes ``amount · coeff`` to
        # *each* quantity its stock carries (CO2 books both CARBON and OXYGEN),
        # instead of all of ``amount`` to one. A 1:1 stock folds ``amount · 1.0``
        # to its single quantity — bit-identical to the pre-P2.1 path. Each
        # quantity has its own accumulator, so inner iteration order is irrelevant;
        # the canonical-order (#15) guarantee comes from the outer stock-id sort.
        for quantity, coeff in stocks[leg.stock].composition.items():
            residual[quantity] = residual.get(quantity, 0.0) + leg.amount * coeff
    return residual


def assert_flow_balanced(
    result: FlowResult,
    stocks: Mapping[StockId, Stock],
    *,
    atol: float = BALANCE_ATOL,
    rtol: float = BALANCE_RTOL,
) -> None:
    """Raise ``ConservationError`` if any asserted quantity fails to balance.

    For each quantity in ``ASSERTED_QUANTITIES`` (every ``Quantity`` — ``ENERGY``
    joined the conserved set in Phase 5), require
    ``abs(residual) <= atol + rtol * scale`` where ``scale`` is the transfer
    magnitude (max ``|leg.amount|`` for that quantity).
    The relative term keeps the check meaningful for the step-8 gate over real,
    varied magnitudes; step-3 synthetic flows (residual ~0) pass on ``atol`` alone.
    """
    residual = per_quantity_residual(result, stocks)
    scale: dict[Quantity, float] = {}
    for leg in result.legs:
        # Same composition fold as the residual: the tolerance denominator for
        # quantity ``q`` is ``max |leg.amount · coeff_q|`` (P2.1). 1:1 folds to
        # ``abs(leg.amount)`` — unchanged.
        for quantity, coeff in stocks[leg.stock].composition.items():
            scale[quantity] = max(scale.get(quantity, 0.0), abs(leg.amount * coeff))
    # Iterate sorted so the first reported failure is deterministic (a frozenset
    # has no defined iteration order); which quantity fails first is cosmetic, but
    # determinism everywhere is cheap and on-brand.
    for quantity in sorted(ASSERTED_QUANTITIES, key=lambda q: q.name):
        r = residual.get(quantity, 0.0)
        tol = atol + rtol * scale.get(quantity, 0.0)
        if abs(r) > tol:
            raise ConservationError(
                f"flow not balanced for {quantity.name}: "
                f"residual {r!r} exceeds tolerance {tol!r}"
            )


def domains_touched(
    result: FlowResult, stocks: Mapping[StockId, Stock]
) -> frozenset[DomainId]:
    """Domains the evaluated ``result`` touches (>1 ⇒ the flow is cross-domain).

    A leg referencing an unknown stock id raises ``KeyError`` (see
    ``per_quantity_residual`` — referential integrity is deferred to the apply
    path, step 5/6).
    """
    return frozenset(stocks[leg.stock].domain for leg in result.legs)

File: rust/crates/simcore/src/flow.rs
//! Flows: structured, atomic, stoichiometric transfers — the Rust port of
//! `simcore.flow`.
//!
//! A [`Flow`] exposes only [`Flow::evaluate`]; legs exist **only after** evaluation
//! against a snapshot, so balance is an *evaluation-time* property of a [`FlowResult`]
//! (per conserved quantity the legs sum to ~0), checked by [`assert_flow_balanced`].
//! Referential integrity (a leg naming a real stock) is the integrator apply path's
//! job, not checked here — a leg on an unknown stock surfaces as [`SimError::Reference`]
//! in [`per_quantity_residual`] / [`assert_flow_balanced`], mirroring Python's
//! `KeyError`.
//!
//! ENERGY is asserted like the mass quantities (it joined the conserved set in
//! Phase 5); [`per_quantity_residual`] reports every quantity present and
//! [`assert_flow_balanced`] iterates all of [`ASSERTED_QUANTITIES`].

use std::collections::{BTreeMap, BTreeSet};

use crate::environment::Environment;
use crate::error::SimError;
use crate::ids::{DomainId, StockId};
use crate::quantities::{Quantity, ASSERTED_QUANTITIES, BALANCE_ATOL, BALANCE_RTOL};
use crate::state::{State, Stock};

/// One stock touched by a flow. `amount` is per dt in the stock's canonical unit:
/// `> 0` deposits, `< 0` withdraws.
#[derive(Debug, Clone, PartialEq)]
pub struct Leg {
    pub stock: StockId,
    pub amount: f64,
}

impl Leg {
    /// Construct a leg, rejecting a non-finite amount (Python `Leg.__post_init__`).
    pub fn new(stock: StockId, amount: f64) -> Result<Leg, SimError> {
        if !amount.is_finite() {
            return Err(SimError::Validation(format!(
                "Leg on {stock:?} amount is not finite: {amount:?}"
            )));
        }
        Ok(Leg { stock, amount })
    }
}

/// The requested transfer from one `evaluate` against a snapshot. At most one leg per
/// stock — a flow nets its own touches into a single leg.
#[derive(Debug, Clone, PartialEq)]
pub struct FlowResult {
    pub legs: Vec<Leg>,
}

impl FlowResult {
    /// Construct a result, rejecting a duplicate leg on any stock (Python
    /// `FlowResult.__post_init__`). An empty `legs` is a valid no-op.
    pub fn new(legs: Vec<Leg>) -> Result<FlowResult, SimError> {
        let mut seen: BTreeSet<&StockId> = BTreeSet::new();
        for leg in &legs {
            if !seen.insert(&leg.stock) {
                return Err(SimError::Validation(format!(
                    "FlowResult has a duplicate leg for stock {:?}; a flow must net \
                     its own touches on a stock into one leg",
                    leg.stock
                )));
            }
        }
        Ok(FlowResult { legs })
    }

    /// The empty (no-op) result.
    pub fn empty() -> FlowResult {
        FlowResult { legs: Vec::new() }
    }
}

/// A pure, deterministic stoichiometric transfer (Python's `Flow` Protocol).
///
/// `evaluate` reads the snapshot/env only, never mutates, and returns the *per-step
/// increment* (`dt·rate`, increment-form — `rate` must be independent of `dt`, which
/// is what lets RK4's ⅙-combine reproduce classical RK4). `priority` is carried for
/// declared-controller policies but is unused under the proportional default;
/// canonical order is always id-sorted, never priority-sorted.
pub trait Flow {
    /// The canonical flow id (ASCII — its sort order is the reduction order).
    fn id(&self) -> &str;

    /// Declared priority (unused under the proportional default; defaults to 0).
    fn priority(&self) -> i64 {
        0
    }

    /// Evaluate the flow against `snapshot`/`env`, returning its per-step legs.
    fn evaluate(
        &self,
        snapshot: &State,
        env: &dyn Environment,
        dt: f64,
    ) -> Result<FlowResult, SimError>;
}

/// Net leg sum per `Quantity` present in `result` — the diagnostic the balance gate
/// composes. Legs are folded in **canonical order (sorted by stock id)**, so the sum
/// is bit-identical regardless of leg construction order (each quantity accumulates
/// in sorted-leg order). A 1:1 stock folds `amount·1.0` to its single quantity.
///
/// A leg naming an unknown stock returns [`SimError::Reference`] (Python `KeyError`).
pub fn per_quantity_residual(
    result: &FlowResult,
    stocks: &BTreeMap<StockId, Stock>,
) -> Result<BTreeMap<Quantity, f64>, SimError> {
    let mut order: Vec<&Leg> = result.legs.iter().collect();
    order.sort_by(|a, b| a.stock.cmp(&b.stock));
    let mut residual: BTreeMap<Quantity, f64> = BTreeMap::new();
    for leg in order {
        let stock = stocks.get(&leg.stock).ok_or_else(|| {
            SimError::Reference(format!("unknown stock {:?} in flow leg", leg.stock))
        })?;
        for (quantity, coeff) in &stock.composition {
            let acc = residual.entry(*quantity).or_insert(0.0);
            *acc += leg.amount * coeff;
        }
    }
    Ok(residual)
}

/// Raise [`SimError::Conservation`] if any asserted quantity fails to balance:
/// `abs(residual) > atol + rtol * scale`, where `scale` is the transfer magnitude
/// (`max |leg.amount·coeff|`) for that quantity. Quantities are checked in name-sorted
/// order so the first reported failure is deterministic.
pub fn assert_flow_balanced(
    result: &FlowResult,
    stocks: &BTreeMap<StockId, Stock>,
    atol: f64,
    rtol: f64,
) -> Result<(), SimError> {
    let residual = per_quantity_residual(result, stocks)?;
    // scale: max |leg.amount·coeff| per quantity (leg order irrelevant — it is a max).
    let mut scale: BTreeMap<Quantity, f64> = BTreeMap::new();
    for leg in &result.legs {
        let stock = stocks.get(&leg.stock).ok_or_else(|| {
            SimError::Reference(format!("unknown stock {:?} in flow leg", leg.stock))
        })?;
        for (quantity, coeff) in &stock.composition {
            let acc = scale.entry(*quantity).or_insert(0.0);
            *acc = acc.max((leg.amount * coeff).abs());
        }
    }
    for quantity in ASSERTED_QUANTITIES {
        let r = residual.get(&quantity).copied().unwrap_or(0.0);
        let tol = atol + rtol * scale.get(&quantity).copied().unwrap_or(0.0);
        if r.abs() > tol {
            return Err(SimError::Conservation(format!(
                "flow not balanced for {}: residual {r:?} exceeds tolerance {tol:?}",
                quantity.name()
            )));
        }
    }
    Ok(())
}

/// The default-tolerance [`assert_flow_balanced`] (using [`BALANCE_ATOL`] /
/// [`BALANCE_RTOL`]).
pub fn assert_flow_balanced_default(
    result: &FlowResult,
    stocks: &BTreeMap<StockId, Stock>,
) -> Result<(), SimError> {
    assert_flow_balanced(result, stocks, BALANCE_ATOL, BALANCE_RTOL)
}

/// The domains the evaluated `result` touches (`> 1` ⇒ cross-domain). A leg naming an
/// unknown stock returns [`SimError::Reference`].
pub fn domains_touched(
    result: &FlowResult,
    stocks: &BTreeMap<StockId, Stock>,
) -> Result<BTreeSet<DomainId>, SimError> {
    let mut domains: BTreeSet<DomainId> = BTreeSet::new();
    for leg in &result.legs {
        let stock = stocks.get(&leg.stock).ok_or_else(|| {
            SimError::Reference(format!("unknown stock {:?} in flow leg", leg.stock))
        })?;
        domains.insert(stock.domain.clone());
    }
    Ok(domains)
}

/// A read-only view of a flow id + its priority — a convenience for tests/callers.
pub fn flow_id(flow: &dyn Flow) -> &str {
    flow.id()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::quantities::StockKind;

    fn co2_stock(id: &str) -> Stock {
        // A gas-phase composition stock: CO₂ = {Carbon:1, Oxygen:2}.
        Stock::new(
            id.to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            10.0,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::from([(Quantity::Carbon, 1.0), (Quantity::Oxygen, 2.0)]),
        )
        .unwrap()
    }

    fn o2_stock(id: &str) -> Stock {
        Stock::new(
            id.to_string(),
            "d".to_string(),
            Quantity::Oxygen,
            "mol".to_string(),
            10.0,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::from([(Quantity::Oxygen, 2.0)]),
        )
        .unwrap()
    }

    #[test]
    fn rejects_duplicate_leg() {
        let e = FlowResult::new(vec![
            Leg::new("a".to_string(), 1.0).unwrap(),
            Leg::new("a".to_string(), -1.0).unwrap(),
        ]);
        assert!(matches!(e, Err(SimError::Validation(_))));
    }

    #[test]
    fn composition_fold_books_both_quantities() {
        // Withdraw 1 CO₂ unit, deposit into another CO₂ pool → balanced for both
        // CARBON and OXYGEN via the composition fold.
        let stocks = BTreeMap::from([
            ("co2.a".to_string(), co2_stock("co2.a")),
            ("co2.b".to_string(), co2_stock("co2.b")),
        ]);
        let result = FlowResult::new(vec![
            Leg::new("co2.a".to_string(), -1.0).unwrap(),
            Leg::new("co2.b".to_string(), 1.0).unwrap(),
        ])
        .unwrap();
        let res = per_quantity_residual(&result, &stocks).unwrap();
        assert_eq!(res.get(&Quantity::Carbon), Some(&0.0));
        assert_eq!(res.get(&Quantity::Oxygen), Some(&0.0));
        assert!(assert_flow_balanced_default(&result, &stocks).is_ok());
    }

    #[test]
    fn imbalanced_flow_is_caught() {
        let stocks = BTreeMap::from([
            ("o2.a".to_string(), o2_stock("o2.a")),
            ("o2.b".to_string(), o2_stock("o2.b")),
        ]);
        // Withdraw 1 unit, deposit 0.5 — OXYGEN residual = 2*(-1)+2*(0.5) = -1.0.
        let result = FlowResult::new(vec![
            Leg::new("o2.a".to_string(), -1.0).unwrap(),
            Leg::new("o2.b".to_string(), 0.5).unwrap(),
        ])
        .unwrap();
        let e = assert_flow_balanced_default(&result, &stocks);
        assert!(matches!(e, Err(SimError::Conservation(_))));
    }

    #[test]
    fn unknown_stock_is_reference_error() {
        let stocks = BTreeMap::from([("o2.a".to_string(), o2_stock("o2.a"))]);
        let result = FlowResult::new(vec![Leg::new("missing".to_string(), 1.0).unwrap()]).unwrap();
        assert!(matches!(
            per_quantity_residual(&result, &stocks),
            Err(SimError::Reference(_))
        ));
    }
}
